### PR TITLE
Remove environment variables in favor of documented config options.

### LINKED ISSUE #822: Remove environment variables.
We currently have a growing number of environment variables that control various parts of the program, acting as a second, parallel configuration system. This means that the config file is not sufficient to determine the behavior of the application. In addition, these variables aren't discoverable like the rest of the configuration system (whose documentation is auto-generated) and we have to manage conflicts between the normal config system and the environment variables.  To fix this, we should remove the environment variables we have so far and convert them to config options, or drop them entirely.

### PR DESCRIPTION
**Pull request by hdevalence**
Closes #822

### CODE REVIEW

**Comment by hdevalence** [comment]
Closes #825

File: zebra-state/src/lib.rs
Comment (hdevalence) [inline zebra-state/src/lib.rs:228]: This test disappears because configs can no longer be invalid.

File: zebrad/src/commands.rs
Comment (hdevalence) [inline zebrad/src/commands.rs:67]: In the new logic we only change the tracing level from `warn` if we're running a server, so this function (which is the inverse of `is_server`) is no longer used.

File: zebrad/src/commands.rs
Comment (hdevalence) [inline zebrad/src/commands.rs:74]: I removed this because it's not strictly speaking a usage note on `ZebradCmd` but about abscissa, and the new tracing initialization logic doesn't have to handle the `None` case differently.

File: zebra-state/src/lib.rs
Comment (teor2345) [reply zebra-state/src/lib.rs:228]: There are other ways of generating a valid config, such as an unreadable directory. But I don't think there's much point in testing sled panics.

File: zebrad/src/application.rs
Comment (teor2345) [inline zebrad/src/application.rs:192]: This change removes support for the verbose logging flag `zebrad -v start`.

I'd like to keep the verbose flag, because it's really useful during development, particularly now we have removed `ZEBRAD_LOG`.

But if you'd like to remove the verbose flag, we should also remove it from the zebrad command-line options, and the flags usage message.

You can see the flags usage message by supplying an unknown flag, for example `zebrad -x`. (They really should show up in `zebrad help` with the subcommands, but that's a separate issue.)

I've looked at the abscissa docs, and I think we might need to implement our own command option parsing and help to disable this option:
https://docs.rs/abscissa_core/0.5.2/abscissa_core/command/trait.Command.html
https://docs.rs/abscissa_core/0.5.2/abscissa_core/struct.Help.html

The help is probably the more important change.

**Review by teor2345 — DISMISSED** [review]
I really like the simpler option handling without env vars.

I had a question about the verbose command-line flag, but I don't want to block on it.

**Comment by oxarbitrage** [comment]
Apart from the verbose flag issue @teor2345 mentioned it is a good change in general. Support.

File: zebrad/src/application.rs
Comment (hdevalence) [reply zebrad/src/application.rs:192]: Yes, we should get rid of this option. I didn't change it because any changes to Abscissa are currently blocked on resolving the existing in-flight changes to Abscissa. We can remove the config option from Abscissa after those changes land, but I don't think we should block these changes on that work.

Comment (oxarbitrage) [reply zebrad/src/application.rs:192]: Opened https://github.com/ZcashFoundation/zebra/issues/831 for this.

**Review by oxarbitrage — APPROVED** [review]

File: zebra-state/src/lib.rs
Comment (yaahc) [inline zebra-state/src/lib.rs:77]: This seems to be a change in behaviour from the current implementation. Do we want to default to using a visible directory in the current directory when we don't have a dirs::cache_dir?

File: zebra-test/src/command.rs
Comment (yaahc) [inline zebra-test/src/command.rs:30]: I don't feel like we should be using serialized toml files to construct configs, I'd much prefer to have a function for creating a `Config` file and one for writing that `Config` to the disk as toml.

File: zebrad/src/application.rs
Comment (yaahc) [reply zebrad/src/application.rs:192]: > Yes, we should get rid of this option. I didn't change it because any changes to Abscissa are currently blocked on resolving the existing in-flight changes to Abscissa. We can remove the config option from Abscissa after those changes land, but I don't think we should block these changes on that work.

I should really finish upstreaming the changes we've made to abscissa...

**Review by yaahc — CHANGES_REQUESTED** [review]

File: zebra-state/src/lib.rs
Comment (hdevalence) [reply zebra-state/src/lib.rs:77]: Yes, I think it's fine.

File: zebra-test/src/command.rs
Comment (hdevalence) [reply zebra-test/src/command.rs:30]: This seemed much simpler, and avoids creating a cyclic dependency between `zebra-test` and `zebrad`.  In the future this could be resolved by moving this `zebrad`-specific test code into the `zebrad` crate, where the `Config` is available.  But I would prefer not to do restructuring of the test code in this PR.

**Review by yaahc — APPROVED** [review]

File: zebra-state/src/lib.rs
Comment (teor2345) [reply zebra-state/src/lib.rs:77]: The setups that don't have a cache directory might not have a home directory (chroots or jails) or it might not be writeable (CI).

So a visible folder in the current directory is a reasonable choice.

File: zebra-test/src/command.rs
Comment (teor2345) [reply zebra-test/src/command.rs:30]: I'm not too concerned about using fixed configs in tests.

Worse case scenario, the tests fail because we change the config format, and we fix the tests.

**Pull request merged**
